[Draft] [ICRC X] Batch canisters call (#220)
dfinity/wg-identity-authentication: topics/icrc_X_multiple_canister_call.md

# ICRC-X: Call batch canisters

[![Status Badge](https://img.shields.io/badge/STATUS-DRAFT-ffcc00.svg)](#)
[![Extension Badge](https://img.shields.io/badge/Extends-ICRC--25-ffcc222.svg)](./icrc_25_signer_interaction_standard.md)
[![Extension Badge](https://img.shields.io/badge/Extends-ICRC--49-ffcc222.svg)](./icrc_49_call_canister.md)

<!-- TOC -->

- [ICRC-X: Call batch canisters](#icrc-x-call-batch-canisters)
  - [Summary](#summary)
  - [Processing](#processing)
    - [Examples](#examples)
      - [Successful Batch Call](#successful-batch-call)
      - [Batch Call with Error](#batch-call-with-error)

## Summary

> This Method can be used by the relying party to request batch call to 3rd party canister executed by the signer using the requested identity. In order to prevent misuse of this method all `icrcX_call_batch_canisters` requests are subject to user approval.

Reviewer: ```suggestion
This method can be used by the relying party to request a batch call to 3rd party canister executed by the signer using the requested identity. In order to prevent misuse of this method all `icrcX_call_batch_canisters` requests are subject to user approval.
```

There is two main parameter for this standard

> 1. `mode` : The mode to execute transaction in sequence or in parallel

Reviewer: Instead of picking between a mode, wouldn't it be more flexible to combine modes?

For example a user might want to ask 2 approvals in parallel before making a call to a swap. To do something like that I'd either nest requests inside requests of for even more flexibility, add a dependency array to each request that indicates which requests must be finished before the request is executed.

An example of optional dependency arrays would look something like this.

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "icrcX_batch_call_canisters",
  "params": {
    "sender": "b7gqo-ulk5n-2kpo7-oalt7-p2kyl-o4j5l-kiuwo-eeybr-dab4l-ur6up-pqe",
    "requests": [
      {
        "id": 1,
        "canisterId": "xhy27-fqaaa-aaaao-a2hlq-ca",
        "method": "approve",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
      },
      {
        "id": 2,
        "canisterId": "xhy27-fqaaa-aaaao-a2hlq-ca",
        "method": "approve",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
      },
      {
        "id": 3,
        "canisterId": "xhy27-fqaaa-aaaao-a2hlq-ca",
        "method": "swap",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
        "waitFor": [1, 2],
      },
      {
        "id": 4,
        "canisterId": "xhy27-fqaaa-aaaao-a2hlq-ca",
        "method": "swap",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
        "waitFor": [3],
      }
    ]
  }
}
```

Requests would execute in this example: `1 & 2 -> 3 -> 4`

Author: but how do we know the dependencies 1 or 2 are success or not in order to execute 3

Author: So from the question in discord 

```
If we really want to implement sequence, we could ask the relying party to pass the expected success response, such as 'ok' (i..e  ICRC 1 or 2 ) or a non-null blockid, for each transaction as parameters to signer, and signer continues to execute the next tx only if the success code is received.
```

Can this possible to do? 
Based on your answer is yes for signer parsing data
But how about relying party passing `expected success response` to the signer like regex or maybe a full payload

2. `requests` : An array of canister call requests

> Unfortunately there is currently no mode to execute all transactions in the batch in one block. The transactions are handled as independent transactions that are executed in parallel or in sequence. However, ICRCX allows signer to receive approval for all transactions in the batch from the user with one approval action. 

Reviewer: Since it's a single approval action and we want to be sure that the user is fully aware and have an overview of what it's approving (e.g. read all consent messages). Should there be a maximum limit of nr of calls?

## Processing
**How it works**
1. Relying party requests a ICRCX batch transaction request to the signer. The request can be either `parallel` mode or `sequence` mode (see more details below). 

2. The signer fetches consent messages and shows a warning to user to approve the batch transaction. 

3. The signer submits canister calls to target canisters:
- mode `parallel`: Execute all requests simultaneously, without waiting for any individual request to complete before starting the next. All transactions will be started. But it is not guaranteed that all transactions will succeed.
- mode `sequence`: 
  - Execute each request one after the other, ensuring that each request is completed before starting the next one. If any transaction fails, the execution of the batch will stop without executing remaining queued transactions. 
  - Evaluate the result of processing the request:

>      1. If the result is a successful response, add the result to the batch response. Proceed with processing the next request in the batch.

Reviewer: As mentioned in the WG call and also Discord, we can't see if the call itself is successful regarding canister logic e.g. success or error result. But we can see (in the certificate) if the call has successfully [received a reply](https://github.com/dfinity/agent-js/blob/main/packages/agent/src/polling/index.ts#L65) from the canister or if something went wrong in between. 

This reply can be any byte array, in practice this is always cbor that will be decoded by the agent-js actor with knowledge of the candid definition of the canister. But even decoded from cbor into js values, there is no clear indicator of success or error, the value that indicates success or error is up to every canisters implementation itself to decide. It could be a boolean, enum or number range that indicates success or error.

So all we can do within this limitation is to stop the chain of calls when:
- No reply has been received
- Something else went wrong in between e.g. IC could not be reached

        
        S ->> U: Display success/failed message
        S ->> RP: Return the batch call response
            
        else Rejected
            U ->> S: Reject request
            S ->> U: Display reject message
            S ->> RP: Error response: Action aborted (3001)
        end
    end
```

### Examples

#### Successful Batch Call

Request

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "icrcX_batch_call_canisters",
  "params": {
    "mode": "sequence",
    "sender": "b7gqo-ulk5n-2kpo7-oalt7-p2kyl-o4j5l-kiuwo-eeybr-dab4l-ur6up-pqe",
    "requests": [
      {
        "id": 1,
        "canisterId": "xhy27-fqaaa-aaaao-a2hlq-ca",
        "method": "transfer",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
      },
      {
        "id": 2,
        "canisterId": "xhy27-fqaaa-aaaao-a2hlq-ca",
        "method": "transfer",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
      }
    ]
  }
}
```

Response

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "responses": [
      {
        "id": 1,
        "result": {

>           "Ok": "16509194"

Reviewer: ```suggestion
          "contentMap": "2dn3p2NhcmdYTkRJREwEbXtuAGwCs7DawwNorYbKgwUBbAP7ygECot6U6wYB2KOMqA19AQMBHVfs7SoKuxOJdX37k7sBJsPALWEo5ayflizuaH0CAADIAWtjYW5pc3Rlcl9pZEoAAAAAAcDR1wEBbmluZ3Jlc3NfZXhwaXJ5GxeNX/65y4YAa21ldGhvZF9uYW1laHRyYW5zZmVyZW5vbmNlUFF4+hAimFhoqkdUcIchz0xscmVxdWVzdF90eXBlZGNhbGxmc2VuZGVyWB1q63Snu+4C5/fpWFu4nq1IpZxCYDEYA8XSPqPfAg==",
          "certificate": "2dn3omR0cmVlgwGDAYIEWCAPzKZJY/emKhi2GGtBrnHh4cdttATd4+9GtJrNCBepb4MBgwJOcmVxdWVzdF9zdGF0dXODAYIEWCCCgynUaonrKCCywghWCSk9BeDqMoI4yf15nxyU/5JZv4MBggRYIDG7WdzQ9sGWI1MpxizUzxubsEBuNkTT94UOZ9USbzNvgwGCBFggawwbTHxnPUzBAUhWBRjk0nzPs2fPpJlaIYtj5AvcX+ODAYIEWCDiFLyaWuMWjtVurCQcSgny/cqfM8S6qrdihVq7nPz1FoMCWCD/8jdeccvqHVYf06Hw7qPXIDNimC1Uyf47VsvgqKpPiIMBgwJFcmVwbHmCA1RESURMAWsCvIoBfcX+0gFxAQAABIMCRnN0YXR1c4IDR3JlcGxpZWSCBFgg7qZngcNt2+B/RuF44W3LRsKWXG6QQg2L6GdZgJ6Nb3+DAYIEWCAx3tU/mhHfX+wDzF003eSJYN8Nebou8rTeGyxr/rUa1YMCRHRpbWWCA0nw9+r88fjXxhdpc2lnbmF0dXJlWDCXNshvwWG1jGViP7ELePGHCThBw9mts45FxIy4gZATkUEsPeJ6y+cjbn2REmB0Soo="
```

As shown [here](https://github.com/dfinity/wg-identity-authentication/blob/main/topics/icrc_49_call_canister.md#example-rpc-response) in ICRC-49, the signer can't and shouldn't decode the response here since it doesn't know the response encoding and the relying party needs the original certificate and contentMap to verify the call has actually been made as requested.

          "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
        }
      },
      {
        "id": 2,
        "params": {
          "canisterId": "xhy27-fqaaa-aaaao-a2hlq-ca",
          "method": "transfer",
          "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
        }
      },
      ,
      {
        "id": 3,
        "params": {
          "canisterId": "xhy27-fqaaa-aaaao-a2hlq-ca",
          "method": "transfer",
          "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
        }
      }
    ]
  }
}
```

Response

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "responses": [
      {
        "id": 1,
        "result": {
          "version": "1",
          "Ok": "16509194"
        }
      },
      {
        "id": 2,
        "result": {
          "Err": {
            "InsufficientFunds": {
              "balance": {
                "e8s": "0"
              }
            }
          }
        }
      },
      {
        "id": 3,
        "error": {

>           "code": 1000,

Reviewer: Would be nice to also have an example where the processing of calls has stopped in between after an error with a call:

- result -> ...
- result -> ...
- error -> code: 4000
- error -> code: 1001
- error -> code: 1001

          "message": "Network error: Unable to establish a connection to the server"
        }
      }
    ]
  }
}
```

## Errors

In addition to the errors defined in [ICRC-25](./icrc_25_signer_interaction_standard.md#errors-3) this standard defines the following errors:

| Code  | Message                                    | Meaning                                                                                          | Data                                                                                |
|-------|--------------------------------------------|--------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------|

> | 10101 | Not processed due to batch request failure | The message was not processed as one of the preceding request in the batch resulted in an error. | (optional) Error details: <ul> <li>`message` (`text`, optional): message</li> </ul> |

Reviewer: Error codes have been [updated in ICRC-49](https://github.com/dfinity/wg-identity-authentication/blob/main/topics/icrc_25_signer_interaction_standard.md#errors-3) a while back, I would recommend to use error code 1001 since it doesn't fit within 2xxx, 3xxx or 4xxx categories.

  "method": "icrcX_batch_call_canisters",
  "params": {
    "sender": "b7gqo-ulk5n-2kpo7-oalt7-p2kyl-o4j5l-kiuwo-eeybr-dab4l-ur6up-pqe",
    "requests": [
      {
        "id": "1",
        "canisterId": "eeddf-fqaaa-aaaao-a2hlq-ca",
        "method": "icrc2_approve",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
      },
      {
        "id": "2",
        "canisterId": "aaabb-fqaaa-aaaao-a2hlq-ca",
        "method": "icrc2_approve",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
      },
      {
        "id": "3",
        "canisterId": "xyzzz-fqaaa-aaaao-a2hlq-ca",
        "method": "swap",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB",
        "waitFor": ["1", "2"]
      },
      ,
      {
        "id": "4",
        "canisterId": "bbbbb-fqaaa-aaaao-a2hlq-ca",
        "method": "bridge_to_eth",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB",
        "waitFor": ["3"],

>         "validateCanister": "xyzzz-fqaaa-aaaao-a2hlq-ca"

Reviewer: Assuming that the validateCanister transform the response from request 4 into success true/false, it doesn't have any affect when defined in the last request since either way no other calls are executed after the last call.

So it should probably be moved to request 3 (maybe also 1 and 2?).

Is there a use case for having a different validateCanister for requests within the same batch? If not it might make more sense to move it up below the sender, this simplifies the signer implementation and enables it to validate multiple responses at once.

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "icrcX_batch_call_canisters",
  "params": {
    "sender": "b7gqo-ulk5n-2kpo7-oalt7-p2kyl-o4j5l-kiuwo-eeybr-dab4l-ur6up-pqe",
    "validate": {
      canisterId: "xyzzz-fqaaa-aaaao-a2hlq-cai"
      method: "validate_batch_approve_swap_bridge"
    },
    "requests": [
      {
        "id": "1",
        "canisterId": "eeddf-fqaaa-aaaao-a2hlq-ca",
        "method": "icrc2_approve",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
      },
      {
        "id": "2",
        "canisterId": "aaabb-fqaaa-aaaao-a2hlq-ca",
        "method": "icrc2_approve",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB"
      },
      {
        "id": "3",
        "canisterId": "xyzzz-fqaaa-aaaao-a2hlq-ca",
        "method": "swap",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB",
        "waitFor": ["1", "2"]
      },
      ,
      {
        "id": "4",
        "canisterId": "bbbbb-fqaaa-aaaao-a2hlq-ca",
        "method": "bridge_to_eth",
        "arg": "RElETARte24AbAKzsNrDA2ithsqDBQFsA/vKAQKi3pTrBgHYo4yoDX0BAwEdV+ztKgq7E4l1ffuTuwEmw8AtYSjlrJ+WLO5ofQIAAMgB",
        "waitFor": ["3"]
      }
    ]
  }
}
```

The `validate_batch_approve_swap_bridge` method of canister `xyzzz-fqaaa-aaaao-a2hlq-cai` would be called twice, first with an array containing the responses from request 1 and 2 and later with an array containing the response from request 3. Keeping latency to a minimum by limiting the number of validation calls to the number of unique `waitFor` combinations.

Author: It's like ICRC21 for response registry canister. Yeah ok, I will talk this with Mason

Author: Is there any flag to not making a new call to validate the response?

If I'm a developer making the swap and bridge, I should do the validation for allowance before executing the swap and bridge. 

we can put `ignore` 
```
 "validate": {
      canisterId: "xyzzz-fqaaa-aaaao-a2hlq-cai"
      method: "validate_batch_approve_swap_bridge"
      ignore: ["3", "4"]
   },
```

or `validate` is optional here

Reviewer: I would argue that validating the last request is never needed, since there is no request after the last request that depends on its outcome.

In case you don't want to validate 3, then there's no reason to make 4 wait for 3 in the first place. You could argue that you'd want to make 4 wait for 3 independent of the outcome of 3 but that is an edge case I wouldn't cover here, instead simply always validate the responses other calls have waited for. Else we end up with more than two behaviors (parallel or sequential based on outcome) which makes things more complicated than needed.

As for making validate optional overall, I think that makes sense but then we'd need to decide what happens if there are waitFor properties defined, either:
- Throw error "validate is required when call depends on each other"
- Expect wallet to fallback to either defining success based on ICRC knowledge or "the call responded at least".

Considering that validation can be implemented without any reliance on devs that implemented the target canisters or wallets implementing ICRC knowledge correctly. I would argue to go for the first option, it guarantees consistent behavior across all signers since the dapp developer itself is in control.

> # ICRC-X: Call batch canisters

Reviewer: ```suggestion
# ICRC-X: Batch Call canisters
```

ICRCX introduces another standard for signer to check if a call request was successfully processed or not.

Currently, Signers do not parse responses from canisters and nor do the responses contain any information about whether the call request was processed successfully. A target casnister response only lets the Signer know that the call request was received and whether any of the following errors occurred: [ICRC-25](./icrc_25_signer_interaction_standard.md#errors-3).

When Signer handles requests that have dependencies and need to be executed in certain order, Signer takes the `contentMap` from the first call request response, and make another call to the target canister request to receive as true or false whether the call request was successful or not. Once Signer validates that the first call request was successfully processed, Signer can execute the second call request that had a dependency on the first.

```
// canister response
{
  "result" : true
}

// signer received payload
{
  // this content map can be fixed
  "contentMap": "2dn3p2NhcmdYTkRscmVxdWVzdF90eXBlZGNhbGxmc2VuZGVyWB1q63Snu+4C5/fpWFu4nq1IpZxCYDEYA8XSPqPfAg==",
  "certificate": "..."
}
```

## Flow

```mermaid

> sequenceDiagram
>     participant RP as Relying Party
>     participant S as Signer
>     participant U as User
>     participant C as Target Canister
>     participant VC as Validate Canister
> 
>     RP ->> S: Request batch canisters call
>     alt Relying party has not been granted the `icrcX_call_canister` permission scope<br>or the request does not comply with scope restrictions
>         S ->> RP: Error response: Permission not granted (3000)
>     else
> 
>         alt If signer want to use consent message from canisters
>             loop for each call of calls in the `icrcX_call_canister`
>                 S ->> C: Get the consent message
>                 C ->> S: Return consent message
>             end
>         end
> 
>         S ->> U: Showing signing message from consent message or blind message
> 
> 
>         alt Approved
>             U ->> S: Approve request
>             Note over S,C: ICRC X execution
> 
>             loop for every sub-array in requests
>                 S ->> C: Submit canister call requests in sub-array in parallel
>                 S ->> S: Wait for the canister call results
> 
>                 alt if validation prop defined
>                     S ->> VC: Call check with validation canister using contentMap
>                     VC ->> S: Return success or failure as bool response (can be new ICRC)
>                 end
> 
>                 alt If the canister called is success
>                     S ->> S: Add call result to batch call response
>                 else If the canister called is failed <br/> because validation failed or execution failed
>                     S ->> S: Add error to batch call response
>                     S ->> S: Add error with code `1001` for <br/> other non-execute request to batch call response
>                 end
>             end
> 
>             Note over S,C: End of the execution
> 
>         S ->> U: Display success/failed message
>         S ->> RP: Return the batch call response
> 
>         else Rejected
>             U ->> S: Reject request
>             S ->> U: Display reject message
>             S ->> RP: Error response: Action aborted (3001)
>         end
>     end

Reviewer: Updated the sequence diagram to use `par` blocks and do an early check if validation is unavailable while there are multiple sub-arrays in requests.

To simplify the sequence diagram, I removed the details regarding adding responses to a list of responses and returning this whole list, this is already clarified in the examples.

```suggestion
sequenceDiagram
    participant RP as Relying Party
    participant S as Signer
    participant U as User
    participant C as Target Canister
    participant VC as Validate Canister

    RP ->> S: Request batch canister call
    alt Relying party has not been granted the `icrcX_call_canister` permission scope<br>or the request does not comply with scope restrictions
        S ->> RP: Error response: Permission not granted (3000)
    else
        alt Validation unavailable and multiple sub-arrays in requests
            S ->> RP: Error response: Validation required (1002)
        end
        par For all requests
            alt Canister supports ICRC-21
                Note over S,C: Follow the ICRC-21 standard
            else Canister does not support ICRC-21 and signer does not support blind signing
                S ->> RP: Error response: No consent message (2001)
                Note over RP,S: Abort interaction, do not proceed further.
            else Canister does not support ICRC-21 and signer supports blind signing
                S ->> U: Display warning and canister call details (canisterId, sender, method, arg)
                Note over S,U: The warning must inform the user that the canister does not support ICRC-21<br/>The arguments should be decoded, otherwise a stronger warning must be displayed
            end
        end
        alt Approved
            U ->> S: Approve request
            loop For every sub-array in requests
                par For each request
                    S ->> C: Submit canister call
                    S ->> S: Wait for the canister call result
                    alt Call failed
                        S ->> U: Display failure message
                        S ->> RP: Error response: Network error (4000) | Generic error 
                    end
                end
                alt Validation available
                    S ->> VC: Send list with all requests and their responses
                    VC ->> S: Return passed or failed
                    alt Failed
                        S ->> U: Display failed message
                        S ->> RP: Error response: Validation failed (1003)
                    end
                end
            end
            S ->> U: Display success message
            S ->> RP: Return the batch call response
        else Rejected
            U ->> S: Reject request
            S ->> U: Display reject message
            S ->> RP: Error response: Action aborted (3001)
        end
    end
```

```

### Examples

#### Successful Batch Call

> Parallel approve request, swap and then call birdge method

Reviewer: ```suggestion
Parallel approve request, swap and then call bridge method
```

          }
        },
        {
          "result": {
            "contentMap": "2dn3p2NhcmdYTkRJREwEbXtuAGwCs7DawwNorYbKgwUBbAP7ygECot6U6wYB2KOMqA19AQMBHVfs7SoKuxOJdX37k7sBJsPALWEo5ayflizuaH0CAADIAWtjYW5pc3Rlcl9pZEoAAAAAAcDR1wEBbmluZ3Jlc3NfZXhwaXJ5GxeNX/65y4YAa21ldGhvZF9uYW1laHRyYW5zZmVyZW5vbmNlUFF4+hAimFhoqkdUcIchz0xscmVxdWVzdF90eXBlZGNhbGxmc2VuZGVyWB1q63Snu+4C5/fpWFu4nq1IpZxCYDEYA8XSPqPfAg==",
            "certificate": "2dn3omR0cmVlgwGDAYIEWCAPzKZJY/emKhi2GGtBrnHh4cdttATd4+9GtJrNCBepb4MBgwJOcmVxdWVzdF9zdGF0dXODAYIEWCCCgynUaonrKCCywghWCSk9BeDqMoI4yf15nxyU/5JZv4MBggRYIDG7WdzQ9sGWI1MpxizUzxubsEBuNkTT94UOZ9USbzNvgwGCBFggawwbTHxnPUzBAUhWBRjk0nzPs2fPpJlaIYtj5AvcX+ODAYIEWCDiFLyaWuMWjtVurCQcSgny/cqfM8S6qrdihVq7nPz1FoMCWCD/8jdeccvqHVYf06Hw7qPXIDNimC1Uyf47VsvgqKpPiIMBgwJFcmVwbHmCA1RESURMAWsCvIoBfcX+0gFxAQAABIMCRnN0YXR1c4IDR3JlcGxpZWSCBFgg7qZngcNt2+B/RuF44W3LRsKWXG6QQg2L6GdZgJ6Nb3+DAYIEWCAx3tU/mhHfX+wDzF003eSJYN8Nebou8rTeGyxr/rUa1YMCRHRpbWWCA0nw9+r88fjXxhdpc2lnbmF0dXJlWDCXNshvwWG1jGViP7ELePGHCThBw9mts45FxIy4gZATkUEsPeJ6y+cjbn2REmB0Soo="
          }
        }
      ],
      [
        {

>           "error": {
>             "code": 1002,
>             "message": "Validation failed"
>           }

Reviewer: Added content map (request args) and certificate (response) to the validation error, this enables the relying party to also verify the validation call besides the canister calls.

```suggestion
          "error": {
            "code": 1002,
            "message": "Validation failed",
            "contentMap": "2dn3p2NhcmdYTkRJREwEbXtuAGwCs7DawwNorYbKgwUBbAP7ygECot6U6wYB2KOMqA19AQMBHVfs7SoKuxOJdX37k7sBJsPALWEo5ayflizuaH0CAADIAWtjYW5pc3Rlcl9pZEoAAAAAAcDR1wEBbmluZ3Jlc3NfZXhwaXJ5GxeNX/65y4YAa21ldGhvZF9uYW1laHRyYW5zZmVyZW5vbmNlUFF4+hAimFhoqkdUcIchz0xscmVxdWVzdF90eXBlZGNhbGxmc2VuZGVyWB1q63Snu+4C5/fpWFu4nq1IpZxCYDEYA8XSPqPfAg==",
            "certificate": "2dn3omR0cmVlgwGDAYIEWCAPzKZJY/emKhi2GGtBrnHh4cdttATd4+9GtJrNCBepb4MBgwJOcmVxdWVzdF9zdGF0dXODAYIEWCCCgynUaonrKCCywghWCSk9BeDqMoI4yf15nxyU/5JZv4MBggRYIDG7WdzQ9sGWI1MpxizUzxubsEBuNkTT94UOZ9USbzNvgwGCBFggawwbTHxnPUzBAUhWBRjk0nzPs2fPpJlaIYtj5AvcX+ODAYIEWCDiFLyaWuMWjtVurCQcSgny/cqfM8S6qrdihVq7nPz1FoMCWCD/8jdeccvqHVYf06Hw7qPXIDNimC1Uyf47VsvgqKpPiIMBgwJFcmVwbHmCA1RESURMAWsCvIoBfcX+0gFxAQAABIMCRnN0YXR1c4IDR3JlcGxpZWSCBFgg7qZngcNt2+B/RuF44W3LRsKWXG6QQg2L6GdZgJ6Nb3+DAYIEWCAx3tU/mhHfX+wDzF003eSJYN8Nebou8rTeGyxr/rUa1YMCRHRpbWWCA0nw9+r88fjXxhdpc2lnbmF0dXJlWDCXNshvwWG1jGViP7ELePGHCThBw9mts45FxIy4gZATkUEsPeJ6y+cjbn2REmB0Soo="
          }
```
